packages/teams-js/test/public/calendar.spec.ts

```diff
@@ -0,0 +1,275 @@
+
+      const openCalendarItemMessage = utils.findMessageByFunc('calendar.openCalendarItem');
+
+      const data = {
+        success: false,
+        error: 'Something went wrong...',
+      };
+
+      utils.respondToMessage(openCalendarItemMessage, data.success, data.error);
+
+      await openCalendarItemPromise.catch(e => expect(e).toMatchObject(new Error('Something went wrong...')));
+    });
+
+    it('should successfully send the openCalendarItem message', async () => {
+      await utils.initializeWithContext('content');
+      utils.setRuntimeConfig({ apiVersion: 1, supports: { calendar: {} } });
+
+      const openCalendarItemPromise = calendar.openCalendarItem(openCalendarItemParams);
+
+      const openCalendarItemMessage = utils.findMessageByFunc('calendar.openCalendarItem');
+
+      const data = {
+        success: true,
+      };
+
+      utils.respondToMessage(openCalendarItemMessage, data.success);
+      await openCalendarItemPromise;
+
+      expect(openCalendarItemMessage).not.toBeNull();
+      expect(openCalendarItemMessage.args.length).toEqual(1);
+      expect(openCalendarItemMessage.args[0]).toStrictEqual(openCalendarItemParams);
+    });
+
+    it('should resolve promise after successfully sending the openCalendarItem message', async () => {
+      await utils.initializeWithContext('content');
+      utils.setRuntimeConfig({ apiVersion: 1, supports: { calendar: {} } });
+
+      const openCalendarItemPromise = calendar.openCalendarItem(openCalendarItemParams);
+
+      const openCalendarItemMessage = utils.findMessageByFunc('calendar.openCalendarItem');
+
+      const data = {
+        success: true,
+      };
+
+      utils.respondToMessage(openCalendarItemMessage, data.success);
+      expect(openCalendarItemPromise).resolves;
+    });
+  });
+
+  describe('composeMeeting', () => {
+    const composeMeetingParams: calendar.ComposeMeetingParams = {};
+
+    it('should not allow calls before initialization', async () => {
+      expect.assertions(1);
+
+      await calendar
+        .composeMeeting(composeMeetingParams)
+        .catch(e => expect(e).toMatchObject(new Error('The library has not yet been initialized')));
+    });
+
+    Object.keys(FrameContexts)
+      .map(k => FrameContexts[k])
+      .forEach(frameContext => {
+        it(`should not allow calls from ${frameContext} context`, async () => {
+          if (frameContext === FrameContexts.content) {
+            return;
+          }
+
+          expect.assertions(1);
+
+          await utils.initializeWithContext(frameContext);
+
+          await calendar
+            .composeMeeting(composeMeetingParams)
+            .catch(e =>
+              expect(e).toMatchObject(
+                new Error(
+                  `This call is only allowed in following contexts: ["content"]. Current context: "${frameContext}".`,
+                ),
+              ),
+            );
+        });
+      });
```

Comment:
I bet this whole thing could be made a helper function that takes in three parameters: the allowed frame context, a function it should call and expect an error, and the string it should test against.

I'm guessing this code only occurs in three places now (two in this file and one in mail?) but even three copies is probably too many and I'm guessing the test should probably be in many other files.

Up to you if you want to fix it as part of this change though; I've just become very familiar with the code which triggered my spidey sense.

Reply:
Yeah, I went back and forth on this. I can definitely make a shared function that does this but I also kind of like having expect statements in actual test files and not hidden in a helper function somewhere cough cough utils.ts

That being said, there might be some way to make a new file for this sort of thing and name it in such a way that it is not mysterious. I will think about it (but probably not tackle it as part of this change)

Reply:
I think in the interest of putting a bow on this PR I am going to leave this as is for now. I think this is a good idea, but I'm not entirely sure where I would want to put these sort of "cross file tests".